src/Compilers/CSharp/Test/Syntax/Parsing/LambdaParameterParsingTests.cs

```diff
@@ -810,6 +810,7 @@ public void TestNullCheckedSingleParamNoSpaces()
                         N(SyntaxKind.Parameter);
                         {
                             N(SyntaxKind.IdentifierToken, "x");
+                            N(SyntaxKind.ExclamationToken);
```

Comment:
do our parsing tests actually validate that tree invariants are preserved?  we should do that so that we catch cases where we are either missing data in the tree, or we have duplicated data.

Note: such a change can come in a followup PR.

Reply:
I believe there is a string equality assert that ensures the tree is correct (bc I kept running into that failing for a bit), but not 100% sure what you're referring to.

Reply:
That's likely what i'm referring to. Note: if that assert fails, we def should never be checking in (to master).  To me, that specific assert is a hard blocker as it means we're not abiding by a core invariant that everything else needs in order to function properly.  

Thanks!

Reply:
To be clear: merging into feature branch: totally ok.  Merging into master with it possible to hit that assert (either inside our outside of tests): Def not ok.  Thanks! :)

Reply:
Yeah all tests would fail if that assert is hit; currently all passing :)

Reply:
Or at least the no-space test would fail.